src/cargo/core/compiler/job_queue.rs

```diff
@@ -884,21 +895,32 @@ impl<'cfg> DrainState<'cfg> {
         }
     }
 
-    fn handle_error(
-        &self,
-        shell: &mut Shell,
-        err_state: &mut Option<anyhow::Error>,
-        new_err: anyhow::Error,
-    ) {
-        if err_state.is_some() {
-            // Already encountered one error.
-            log::warn!("{:?}", new_err);
-        } else if !self.active.is_empty() {
-            crate::display_error(&new_err, shell);
-            drop(shell.warn("build failed, waiting for other jobs to finish..."));
-            *err_state = Some(anyhow::format_err!("build failed"));
-        } else {
-            *err_state = Some(new_err);
+    fn handle_error(&self, shell: &mut Shell, err_state: &mut DrainErrors, new_err: anyhow::Error) {
+        match err_state {
+            DrainErrors::FailFast { first_error } => {
+                if first_error.is_some() {
+                    // Already encountered one error.
+                    log::warn!("{:?}", new_err);
+                } else if !self.active.is_empty() {
+                    crate::display_error(&new_err, shell);
+                    drop(shell.warn("build failed, waiting for other jobs to finish..."));
+                    *first_error = Some(anyhow::format_err!("build failed"));
+                } else {
+                    *first_error = Some(new_err);
```

Comment:
I keep getting somewhat confused when I think about this personally, and while I realize that this is preserving what's already happening here I think it might be best to restructure this differently. I believe the original intention of this code was, when possible, to return upwards the error we encountered during compilation if only one was encountered. That being said, however, I don't believe we can reliably return upwards an error so I think it would be better to unconditionally, for any errors encountered during the build, to use `crate::display_error` and print them here. That way there's much less juggling of trying to preserve some "main" error to return while also ensuring that all other errors are printed.

With a refactoring like that I think it might make sense to change `DrainErrors` from an enum to a struct with a `keep_going` flag or similar since the only change in behavior would be whether or not more jobs are spawned but otherwise errors are handled consistently between `--keep-going`-vs-not